File: vision/detector/warp_death_tracker.py
"""Warp/death tracker: detects Up+A warps and deaths from gameplay gaps.

Tracks non-gameplay gaps (consecutive non-subscreen, non-gameplay frames)
and detects when gameplay resumes at a known reset position (overworld start
or dungeon entrance). Also handles the CSR screen-based fallback detection
and the hearts-zero streak guard against false deaths.
"""

from typing import Callable


class WarpDeathTracker:
    """Detects Up+A warps and deaths via position-reset and CSR patterns.

    Public attributes overworld_start and dungeon_entrances are read by
    the coordinator for Rule 10 (map adjacency checks).
    """

    def __init__(
        self,
        any_roads: set[int] | None = None,
    ):
        self.overworld_start: int = 0
        self.dungeon_entrances: dict[int, int] = {}

        # any_roads is kept for future use (Rule 10 in coordinator uses it)
        self.any_roads: set[int] = any_roads or set()

        # Hearts-zero streak tracking
        self._last_gameplay_hearts: int = 0
        self._zero_hearts_streak: int = 0

        # Non-gameplay gap tracking
        self._non_gameplay_gap: int = 0
        self._last_gameplay_position: int = 0
        self._last_gameplay_screen: str = ''
        self._warp_detected_this_gap: bool = False

    # ------------------------------------------------------------------
    # Public interface
    # ------------------------------------------------------------------

    def process_frame(
        self,
        screen_type: str,
        dungeon_level: int,
        hearts_current: int,
        hearts_max: int,
        map_position: int,
        prev_screen_type: str,
        prev_hearts_max: int,
        gameplay_started: bool,
        game_completed: bool,
        game_events: list[dict],
        frame_number: int,
        dungeon_exit_exiting_d9: bool = False,
    ) -> list[dict]:
        """Process one frame. Returns list of game event dicts (death, up_a_warp).

        Also updates overworld_start, dungeon_entrances, and internal gap/
        streak state in place.
        """
        events: list[dict] = []

        # ─── Save pre-update hearts for death vs Up+A determination ───
        # Must capture BEFORE the hearts-zero streak update below.
        pre_gap_hearts = self._last_gameplay_hearts

        # ─── Hearts-zero streak tracking ───
        # Require 4 consecutive gameplay frames with hearts=0 AND consistent
        # hearts_max to confirm Link actually died (not a misclassified frame).
        if screen_type in ('overworld', 'dungeon', 'cave'):
            if hearts_current > 0:
                self._last_gameplay_hearts = hearts_current
                self._zero_hearts_streak = 0
            elif prev_hearts_max > 0 and hearts_max >= prev_hearts_max:
                # hearts_max is consistent — HUD was present, 0 is real
                self._zero_hearts_streak += 1
                if self._zero_hearts_streak >= 4:
                    self._last_gameplay_hearts = 0
            # else: hearts_max dropped (transition frame with default values) — ignore

        # ─── Position-reset warp/death detection ───
        if screen_type in ('overworld', 'dungeon', 'cave'):
            if (
                self._non_gameplay_gap >= 4
                and gameplay_started
                and not game_completed
                and not self._warp_detected_this_gap
            ):
                new_pos = map_position
                is_reset = False

                if (
                    screen_type == 'overworld'
                    and self.overworld_start > 0
                    and new_pos == self.overworld_start
                ):
                    is_reset = True
                elif screen_type == 'dungeon' and dungeon_level > 0:
                    entrance = self.dungeon_entrances.get(dungeon_level, 0)
                    if (
                        entrance > 0
                        and new_pos == entrance
                        and self._last_gameplay_screen == 'dungeon'
                    ):
                        is_reset = True

                # Don't fire if triforce was just inferred this frame
                triforce_just_inferred = (
                    game_events
                    and game_events[-1].get('event') == 'triforce_inferred'
                    and game_events[-1].get('frame') == frame_number
                )

                if is_reset and not triforce_just_inferred:
                    self._warp_detected_this_gap = True
                    if pre_gap_hearts == 0:
                        events.append({
                            'frame': frame_number,
                            'event': 'death',
                            'description': (
                                f'Link died (respawned at reset position '
                                f'after {self._non_gameplay_gap} frame gap)'
                            ),
                            'dungeon_level': dungeon_level,
                        })
                    else:
                        events.append({
                            'frame': frame_number,
                            'event': 'up_a_warp',
                            'description': (
                                f'Up+A warp (hearts {pre_gap_hearts}, '
                                f'reset after {self._non_gameplay_gap} frame gap)'
                            ),
                            'dungeon_level': dungeon_level,
                        })

        # ─── CSR-based death/warp detection ───
        if (
            screen_type == 'death'
            and prev_screen_type != 'death'
            and not game_completed
            and gameplay_started
            and not self._warp_detected_this_gap
            and not dungeon_exit_exiting_d9
        ):
            self._warp_detected_this_gap = True
            if self._last_gameplay_hearts == 0:
                events.append({
                    'frame': frame_number,
                    'event': 'death',
                    'description': (
                        'Link died (hearts reached 0, CSR screen detected)'
                    ),
                    'dungeon_level': dungeon_level,
                })
            else:
                events.append({
                    'frame': frame_number,
                    'event': 'up_a_warp',
                    'description': (
                        f'Up+A warp (hearts were {self._last_gameplay_hearts}, '
                        f'CSR screen detected)'
                    ),
                    'dungeon_level': dungeon_level,
                })

        # ─── Update start/entrance positions ───
        # Must happen AFTER position-reset detection (uses old values)
        # and BEFORE the coordinator's Rule 10 adjacency check (reads new values).
        if map_position > 0:
            if screen_type == 'overworld' and self.overworld_start == 0:
                self.overworld_start = map_position
            if (
                screen_type == 'dungeon'
                and dungeon_level > 0
                and dungeon_level not in self.dungeon_entrances
            ):
                self.dungeon_entrances[dungeon_level] = map_position

        # ─── Update non-gameplay gap and last gameplay state ───
        if screen_type in ('overworld', 'dungeon', 'cave'):
            self._non_gameplay_gap = 0
            self._warp_detected_this_gap = False
            self._last_gameplay_position = map_position
            self._last_gameplay_screen = screen_type
        elif screen_type != 'subscreen':
            self._non_gameplay_gap += 1

        return events

    def reset(self) -> None:
        """Clear all state."""
        self.overworld_start = 0
        self.dungeon_entrances.clear()
        self._last_gameplay_hearts = 0
        self._zero_hearts_streak = 0
        self._non_gameplay_gap = 0
        self._last_gameplay_position = 0
        self._last_gameplay_screen = ''
        self._warp_detected_this_gap = False

File: server/src/vision/WarpDeathTracker.ts
import type { StableGameState, GameEvent, GameEventType } from './types.js';

const NON_GAMEPLAY_GAP_THRESHOLD = 4;
const ZERO_HEARTS_STREAK_THRESHOLD = 4;

const GAMEPLAY_SCREENS = new Set(['overworld', 'dungeon', 'cave']);

export class WarpDeathTracker {
  private racerId: string;
  private lastGameplayHearts = 0;
  private zeroHeartsStreak = 0;
  private nonGameplayGap = 0;
  private lastGameplayPosition = -1;
  private lastGameplayScreen = '';
  private warpDetectedThisGap = false;
  private overworldStart = 0;
  private dungeonEntrances = new Map<number, number>();
  private gameplayStarted = false;
  private gameplayStreak = 0;

  constructor(racerId: string) {
    this.racerId = racerId;
  }

  registerStart(position: number): void {
    if (this.overworldStart === 0) this.overworldStart = position;
  }

  registerDungeonEntrance(dungeonLevel: number, position: number): void {
    if (!this.dungeonEntrances.has(dungeonLevel)) {
      this.dungeonEntrances.set(dungeonLevel, position);
    }
  }

  update(state: StableGameState, timestamp: number, frameNumber: number,
         events: GameEvent[], gameCompleted = false): void {
    const isGameplay = GAMEPLAY_SCREENS.has(state.screenType);

    if (isGameplay) {
      this.gameplayStreak++;
      if (this.gameplayStreak >= 120 && !this.gameplayStarted) {
        this.gameplayStarted = true;
      }
    } else if (state.screenType === 'title') {
      this.gameplayStreak = 0;
    }

    if (isGameplay) {
      if (this.nonGameplayGap >= NON_GAMEPLAY_GAP_THRESHOLD
          && this.gameplayStarted && !gameCompleted && !this.warpDetectedThisGap) {
        this._checkPositionReset(state, timestamp, frameNumber, events);
      }

      this.nonGameplayGap = 0;
      this.warpDetectedThisGap = false;
      this.lastGameplayPosition = state.mapPosition;
      this.lastGameplayScreen = state.screenType;
    } else if (state.screenType !== 'subscreen') {
      this.nonGameplayGap++;

      if (state.screenType === 'death' && this.lastGameplayScreen !== ''
          && !gameCompleted && this.gameplayStarted && !this.warpDetectedThisGap) {
        this.warpDetectedThisGap = true;
        if (this.lastGameplayHearts === 0) {
          this._emit(events, 'death', timestamp, frameNumber, 'Link died');
        } else {
          this._emit(events, 'up_a_warp', timestamp, frameNumber,
            `Up+A warp (hearts were ${this.lastGameplayHearts})`);
        }
      }
    }

    if (isGameplay) {
      if (state.heartsCurrentStable > 0) {
        this.lastGameplayHearts = state.heartsCurrentStable;
        this.zeroHeartsStreak = 0;
      } else if (state.heartsMaxStable > 0) {
        this.zeroHeartsStreak++;
        if (this.zeroHeartsStreak >= ZERO_HEARTS_STREAK_THRESHOLD) {
          this.lastGameplayHearts = 0;
        }
      }
    }
  }

  private _checkPositionReset(state: StableGameState, ts: number, fn: number,
                               events: GameEvent[]): void {
    let isReset = false;

    if (state.screenType === 'overworld' && this.overworldStart > 0
        && state.mapPosition === this.overworldStart) {
      isReset = true;
    }

    if (state.screenType === 'dungeon' && state.dungeonLevel > 0) {
      const entrance = this.dungeonEntrances.get(state.dungeonLevel);
      if (entrance !== undefined && entrance > 0
          && state.mapPosition === entrance
          && this.lastGameplayScreen === 'dungeon') {
        isReset = true;
      }
    }

    if (isReset) {
      this.warpDetectedThisGap = true;
      if (this.lastGameplayHearts === 0) {
        this._emit(events, 'death', ts, fn, 'Link died (position reset)');
      } else {
        this._emit(events, 'up_a_warp', ts, fn,
          `Up+A warp — returned to start (hearts ${this.lastGameplayHearts})`);
      }
    }
  }

  private _emit(events: GameEvent[], type: GameEventType, ts: number, fn: number,
                desc: string): void {
    events.push({
      type, racerId: this.racerId, timestamp: ts,
      frameNumber: fn, priority: type === 'death' ? 'high' : 'low',
      description: desc,
    });
  }

  reset(): void {
    this.lastGameplayHearts = 0;
    this.zeroHeartsStreak = 0;
    this.nonGameplayGap = 0;
    this.lastGameplayPosition = -1;
    this.lastGameplayScreen = '';
    this.warpDetectedThisGap = false;
    this.overworldStart = 0;
    this.dungeonEntrances.clear();
    this.gameplayStarted = false;
    this.gameplayStreak = 0;
  }
}
